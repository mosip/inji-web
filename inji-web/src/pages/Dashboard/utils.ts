export const convertStringIntoPascalCase = (text: string | null) => {
    if (text != null) {
        return text
            .toLocaleLowerCase()
            .split(" ")
            .map((word) => word.charAt(0).toUpperCase() + word.slice(1))
            .join(" ");
    }
};
