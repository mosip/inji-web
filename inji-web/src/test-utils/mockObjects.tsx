import { IssuerConfigurationObject } from '../types/data';
import { LandingPageWrapperProps } from '../components/Common/LandingPageWrapper';
import { NavBarProps } from '../types/components';
export const mockIssuerDisplayArrayObject = {
    name: "Name",
    language: "en",
    locale: "en",
    logo: {
        url: "https://url.com",
        alt_text: "alt text of the url"
    },
    title: "Title",
    description: "Description",
};

export const mockCredentialTypeDisplayArrayObject = {
    name: "Name",
    locale: "en",
    logo: "https://url.com",
};

export const mockCredentials: IssuerConfigurationObject = {
    credentials_supported: [
        {
            name: "InsuranceCredential",
            scope: "mosip_vc_ldp",
            display: [
                {
                    name: "Health Insurance",
                    locale: "en",
                    logo: "https://url.com"
                }
            ]
        },
        {
            name: "AnotherCredential",
            scope: "mosip_ldp_vc",
            display: [
                {
                    name: "Another Credential",
                    locale: "en",
                    logo: "https://url.com"
                }
            ]
        }
    ],
    "authorization_endpoint": "https://env.net/authorize",
    "grant_types_supported": ["authorization_code"]
};


export const mockIssuerObject = {
        name: 'Test Issuer',
        desc: 'Test Description',
        protocol: 'OTP' as 'OTP', // Explicitly set the type to 'OTP' or 'OpenId4VCI'
        issuer_id: 'test-issuer',
        authorization_endpoint: 'https://auth.test.com',
        credentials_endpoint: 'https://credentials.test.com',
        display: [{
            name: "Name",
            language: "en",
            locale: "en",
            logo: {
                url: "https://url.com",
                alt_text: "alt text of the url"
            },
            title: "Title",
            description: "Description",
        }],
        client_id: 'test-client-id',
        redirect_uri:'test-redirect-uri',
        token_endpoint:'test-token_endpoint',
        proxy_token_endpoint:'test-proxy_token_endpoint',
        client_alias:'',
        ovp_qr_enabled: true,
        scopes_supported: ['scope1', 'scope2']
}

export const mockIssuerObjectList =  [
    {
        name: 'Issuer 1',
        desc: 'Description 1',
        protocol: 'OpenId4VCI',
        issuer_id: 'issuer1',
        authorization_endpoint: 'https://issuer1.com/auth',
        credentials_endpoint: 'https://issuer1.com/credentials',
        display: [
            {
                name: 'Issuer 1',
                language: 'en',
                locale: 'en-US',
                logo: { url: 'https://issuer1.com/logo.png', alt_text: 'Issuer 1 Logo' },
                title: 'Issuer 1 Title',
                description: 'Issuer 1 Description',
            },
        ],
        client_id: 'client1',
        redirect_uri: 'https://issuer1.com/redirect',
        token_endpoint: 'https://issuer1.com/token',
        proxy_token_endpoint: 'https://issuer1.com/proxy-token',
        client_alias: 'client1-alias',
        ovp_qr_enabled: true,
        scopes_supported: ['openid', 'profile'],
    },
    {
        name: 'Issuer 2',
        desc: 'Description 2',
        protocol: 'OpenId4VCI',
        issuer_id: 'issuer2',
        authorization_endpoint: 'https://issuer2.com/auth',
        credentials_endpoint: 'https://issuer2.com/credentials',
        display: [
            {
                name: 'Issuer 2',
                language: 'en',
                locale: 'en-US',
                logo: { url: 'https://issuer2.com/logo.png', alt_text: 'Issuer 2 Logo' },
                title: 'Issuer 2 Title',
                description: 'Issuer 2 Description',
            },
        ],
        client_id: 'client2',
        redirect_uri: 'https://issuer2.com/redirect',
        token_endpoint: 'https://issuer2.com/token',
        proxy_token_endpoint: 'https://issuer2.com/proxy-token',
        client_alias: 'client2-alias',
        ovp_qr_enabled: false,
        scopes_supported: ['openid', 'profile'],
    },
]

export const mockLandingPageWrapperProps: LandingPageWrapperProps = {
    icon: <div data-testid="Test-Icon">Icon</div>,
    title: "Test Title",
    subTitle: "Test SubTitle",
    gotoHome: true,
};

export const mockNavBarProps: NavBarProps = {
    title: 'Test Title',
    link: '/test-link',
    search: true,
};