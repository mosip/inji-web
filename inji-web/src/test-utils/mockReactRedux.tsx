jest.mock('react-redux', () => {
    const ActualReactRedux = jest.requireActual('react-redux');
    return {
      ...ActualReactRedux,
      useSelector: jest.fn(),
      useDispatch: jest.fn(),
    };
  });
  
  // Helper to set the mock implementation of useSelector:
  export const setMockUseSelectorState = (state: any) => {
    const { useSelector } = require('react-redux') as { useSelector: jest.Mock };
    useSelector.mockImplementation((selector: any) => selector(state));
  };
  
  export const setMockUseDispatchReturnValue = (dispatchFn: jest.Mock) => {
    const { useDispatch } = require('react-redux') as { useDispatch: jest.Mock };
    (useDispatch as jest.Mock).mockReturnValue(dispatchFn);
};