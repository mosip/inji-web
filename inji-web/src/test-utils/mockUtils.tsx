import React, { ReactElement } from 'react';
import { render, RenderOptions } from '@testing-library/react';
import { Provider } from 'react-redux';
import { BrowserRouter, Route, BrowserRouter as Router, Routes } from 'react-router-dom';
import { reduxStore } from '../redux/reduxStore';

// Mock for storage module
export const mockStorageModule = () => {
    jest.mock('../utils/storage.ts', () => ({
        storage: {
            getItem: jest.fn(),
            setItem: jest.fn(),
            SESSION_INFO: 'SESSION_INFO',
            SELECTED_LANGUAGE: 'selectedLanguage'
        }
    }));
};

// Mock for localStorage
export const mockLocalStorage = () => {
    let store: { [key: string]: string } = {};
  
    const localStorageMock = {
      getItem: jest.fn((key: string) => store[key] || null),
      setItem: jest.fn((key: string, value: string) => {
        store[key] = value;
      })
    };
  
    Object.defineProperty(window, 'localStorage', { 
      value: localStorageMock,
      writable: true
    });
  
    return localStorageMock;
  };

// API Mock
export const mockApi = {
    authorizationRedirectionUrl: 'http://mockurl.com'
};

// Crypto Mock
export const mockCrypto = {
    getRandomValues: (array: Uint32Array) => {
        for (let i = 0; i < array.length; i++) {
            array[i] = 4294967296;
        }
        return array;
    },
    subtle: {} as SubtleCrypto,
    randomUUID: () => '123e4567-e89b-12d3-a456-426614174000'
} as Crypto;

export const mockUseTranslation = () => {
    jest.mock('react-i18next', () => ({
        useTranslation: () => ({
            t: (key: string) => key,
        }),
    }));
};

export const mockUseNavigate = () => {
    const mockNavigate = jest.fn();
    jest.mock('react-router-dom', () => ({
        ...jest.requireActual('react-router-dom'),
        useNavigate: mockNavigate,
    }));
};

export const mockUseSelector = () => {
    jest.mock('react-redux', () => ({
        ...jest.requireActual('react-redux'),
        useSelector: jest.fn(),
    }));
};

export const setMockUseSelectorState = (state: any) => {
    const useSelectorMock = require('react-redux').useSelector;
    useSelectorMock.mockImplementation((selector: any) => selector(state));
};

export const mockUseGetObjectForCurrentLanguage = () => {
    jest.mock('../utils/i18n', () => ({
        getObjectForCurrentLanguage: jest.fn(),
    }));
};

export const wrapUnderRouter = (children: React.ReactNode) => {
    return <Router>{children}</Router>;
};

export const mockUseSearchCredentials = () => {
    jest.mock('../components/Credentials/SearchCredential', () => ({
        SearchCredential: () => <></>,
    }));
 };

export const mockUseDispatch = () => {
    jest.mock('react-redux', () => ({
        ...jest.requireActual('react-redux'),
        useDispatch: jest.fn(),
    }));
};
              
export const mockUseSpinningLoader = () => {
    jest.mock('../components/Common/SpinningLoader', () => ({
        SpinningLoader: () => <></>,
    }));
};

export const mockUseLanguageSelector = () => {
    jest.mock('../components/Common/LanguageSelector', () => ({
        LanguageSelector: () => <></>,
    }));
};

interface RenderWithProviderOptions extends Omit<RenderOptions, 'queries'> {}

export const renderWithProvider = (element: ReactElement, options?: RenderWithProviderOptions) => {
    return render(
        <Provider store={reduxStore}>
            <Router>
                {element}
            </Router>
        </Provider>,
        options
    );
};
    
export const mockUseParams = ()=>{
    jest.mock('react-router-dom', () => ({
        ...jest.requireActual('react-router-dom'),
        useParams: () => ({ issuerId: 'test-issuer-id' }),
      }));   
};

export const mockUseapiObject = () =>{
    jest.mock('../utils/api', () => ({
        api: {
          mimotoHost: 'http://mocked-api-host',
        },
      }));
};

export const mockUseFetch = () =>{
    jest.mock('../hooks/useFetch', () => {
        const RequestStatus = {
          LOADING: 'LOADING',
          DONE: 'DONE',
          ERROR: 'ERROR',
        };
        return {
          useFetch: () => jest.fn(),
          RequestStatus,
        };
      });
};

export const mockUseToast = () =>{
    jest.mock('react-toastify', () => ({
      toast: {
        error: jest.fn(),
      },
    }));
};

export const mockusemisc = ()=>{
    jest.mock('../utils/misc', () => ({
        downloadCredentialPDF: jest.fn(),
        getErrorObject: jest.fn(),
        getTokenRequestBody: jest.fn(),
      }));
};

export const mockusei18n = ()=>{
    jest.mock('react-i18next', () => ({
        useTranslation: () => ({
          t: (key: string) => key,
        }),
        initReactI18next: {
          type: '3rdParty',
          init: jest.fn(),
        },
      }));

};

export const mockWindowLocation = (url: string) => {
    const location = new URL(url) as unknown as Location;
    Object.defineProperty(window, 'location', {
      value: location,
      writable: true,
    });
  };

export const renderWithRouter = (Element: React.ReactElement, { route = '/' } = {}) => {
    window.history.pushState({}, 'Test page', route);
    return render(
      <BrowserRouter>
        <Provider store={reduxStore}>
          <Routes>
            <Route path="*" element={Element} />
          </Routes>
        </Provider>
      </BrowserRouter>
    );
  };

