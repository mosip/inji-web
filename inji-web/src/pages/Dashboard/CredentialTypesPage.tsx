import React, {useEffect, useState} from 'react';
import {useLocation, useNavigate, useParams} from 'react-router-dom';
import {RequestStatus, useFetch} from '../../hooks/useFetch';
// import {NavBar} from '../components/Common/NavBar';
import {CredentialList} from '../../components/Credentials/CredentialList';
import {useDispatch, useSelector} from 'react-redux';
import {storeSelectedIssuer} from '../../redux/reducers/issuersReducer';
import {
    storeCredentials,
    storeFilteredCredentials
} from '../../redux/reducers/credentialsReducer';
import {api} from '../../utils/api';
import {useTranslation} from 'react-i18next';
import {toast} from 'react-toastify';

import {
    ApiRequest,
    IssuerWellknownDisplayArrayObject,
    IssuerObject
} from '../../types/data';
import {getIssuerDisplayObjectForCurrentLanguage} from '../../utils/i18n';
import {RootState} from '../../types/redux';
import {isObjectEmpty} from '../../utils/misc';
import {SearchCredential} from '../../components/Credentials/SearchCredential';
import {CredentialTypesPageProps} from '../../components/Dashboard/types';
import {navigateToDashboardHome} from './utils';
import { NavBackArrowButton } from './NavBackArrowButton';

export const CredentialTypesPage: React.FC<CredentialTypesPageProps> = ({
    backUrl
}) => {
    const {state, fetchRequest} = useFetch();
    const params = useParams<CredentialParamProps>();
    const dispatch = useDispatch();
    const {t} = useTranslation(['CredentialsPage', 'Dashboard']);
    const language = useSelector((state: RootState) => state.common.language);
    let displayObject = {} as IssuerWellknownDisplayArrayObject;
    let [selectedIssuer, setSelectedIssuer] = useState({} as IssuerObject);
    if (!isObjectEmpty(selectedIssuer)) {
        displayObject = getIssuerDisplayObjectForCurrentLanguage(
            selectedIssuer.display,
            language
        );
    }
    const navigate = useNavigate();

    useEffect(() => {
        const fetchCall = async () => {
            let apiRequest: ApiRequest = api.fetchSpecificIssuer;
            let response = await fetchRequest(
                apiRequest.url(params.issuerId ?? ''),
                apiRequest.methodType,
                apiRequest.headers()
            );
            dispatch(storeSelectedIssuer(response?.response));
            setSelectedIssuer(response?.response);

            apiRequest = api.fetchIssuersConfiguration;
            response = await fetchRequest(
                apiRequest.url(params.issuerId ?? ''),
                apiRequest.methodType,
                apiRequest.headers()
            );

            dispatch(storeFilteredCredentials(response?.response));
            dispatch(storeCredentials(response?.response));
        };
        fetchCall();
    }, []);

    if (state === RequestStatus.ERROR) {
        toast.error(t('errorContent'));
    }

    const location = useLocation();
    const previousPagePath = location.state?.from;

    const handleBackClick = () => {
        if (backUrl) {
            navigate(backUrl); // Navigate to the URL sent by the parent
        } else if (previousPagePath) {
            navigate(previousPagePath); // Navigate to the previous link in history
        } else {
            navigateToDashboardHome(navigate); // Navigate to homepage if opened directly
        }
    };

    return (
        <div
            data-testid={'Credential-Types-Page-Container'}
            className="container mx-auto sm:px-2 md:px-4 lg:px-6 py-6 flex flex-col srelative gap-4 sm:gap-6 md:gap-10 lg:gap-12 ml-4  sm:ml-0"
        >
            <div className="flex flex-col lg:flex-row justify-between items-start gap-4 sm:gap-2">
                <div className="flex items-start">
                    <div className="flex items-start">
                        <NavBackArrowButton handleBackClick={handleBackClick} />
                    </div>
                    <div className="flex flex-col items-start">
                        <span
                            data-testid={'Stored-Credentials'}
                            className="text-2xl font-medium"
                        >
                            {displayObject?.name}
                        </span>
                        <button
                            data-testid={'Home'}
                            className="text-xs sm:text-sm text-[#5B03AD] cursor-pointer"
                            onClick={() => navigateToDashboardHome(navigate)}
                        >
                            {t('Dashboard:Home.title')}
                        </button>
                    </div>
                </div>
                <div>
                    <SearchCredential
                        issuerContainerBorderRadius={'rounded-md'}
                    />
                </div>
            </div>

            <div className="bg-white rounded-lg shadow-iw-emptyDocuments flex flex-col sm:flex-row justify-between items-start p-4 sm:p-6">
                <div
                    data-testid="Credential-List-Container"
                    className="container mx-auto"
                >
                    <CredentialList state={state} />
                </div>
            </div>
        </div>
    );
};
