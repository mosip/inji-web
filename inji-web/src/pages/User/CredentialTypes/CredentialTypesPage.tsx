import React, {useEffect, useState} from 'react';
import {useLocation, useNavigate, useParams} from 'react-router-dom';
import {RequestStatus, useFetch} from '../../../hooks/useFetch';
import {useDispatch, useSelector} from 'react-redux';
import {storeSelectedIssuer} from '../../../redux/reducers/issuersReducer';
import {storeCredentials, storeFilteredCredentials} from '../../../redux/reducers/credentialsReducer';
import {api} from '../../../utils/api';
import {ApiRequest, IssuerObject, IssuerWellknownDisplayArrayObject} from '../../../types/data';
import {getIssuerDisplayObjectForCurrentLanguage} from '../../../utils/i18n';
import {RootState} from '../../../types/redux';
import {isObjectEmpty} from '../../../utils/misc';
import {navigateToUserHome} from "../../../utils/navigationUtils";
import {CredentialTypesPageStyles} from "./CredentialTypesPageStyles";
import {useDownloadSessionDetails} from "../../../hooks/userDownloadSessionDetails";
import {CredentialTypesPageContent} from "../../../components/User/CredentialTypes/CredentialTypesPageContent.tsx";
import {Header} from "../../../components/User/CredentialTypes/Header.tsx";
import {ROUTES} from "../../../utils/constants";

type CredentialTypesPageProps = {
    backUrl?: string;
};

export const CredentialTypesPage: React.FC<CredentialTypesPageProps> = ({
                                                                            backUrl
                                                                        }) => {
    const {state, fetchRequest} = useFetch();
    const params = useParams<CredentialParamProps>();
    const dispatch = useDispatch();
    const language = useSelector((state: RootState) => state.common.language);
    let displayObject = {} as IssuerWellknownDisplayArrayObject;
    let [selectedIssuer, setSelectedIssuer] = useState({} as IssuerObject);
    if (!isObjectEmpty(selectedIssuer)) {
        displayObject = getIssuerDisplayObjectForCurrentLanguage(
            selectedIssuer.display,
            language
        );
    }
    const navigate = useNavigate();
    const location = useLocation();
    const {
        downloadInProgressSessions,
        currentSessionDownloadId,
        setCurrentSessionDownloadId,
        setLatestDownloadedSessionId
    } = useDownloadSessionDetails();

    const [downloadStatus, setDownloadStatus] = useState<RequestStatus | null>(null);

    useEffect(() => {
        const status = currentSessionDownloadId ? downloadInProgressSessions[currentSessionDownloadId]?.downloadStatus : null;
        setDownloadStatus(status);

    }, [currentSessionDownloadId, downloadInProgressSessions]);

    useEffect(() => {
        return (() => {
            setCurrentSessionDownloadId(null);
            setLatestDownloadedSessionId(null);
        })
    }, []);

    useEffect(() => {
        if (downloadStatus === RequestStatus.DONE) {
            navigate(ROUTES.CREDENTIALS)
        }
    }, [downloadStatus])

    useEffect(() => {
        if (state === RequestStatus.ERROR) {
            setDownloadStatus(RequestStatus.ERROR);
        }
    }, [state])

    useEffect(() => {
        const fetchCall = async () => {
            let apiRequest: ApiRequest = api.fetchSpecificIssuer;
            let response = await fetchRequest(
                apiRequest.url(params.issuerId ?? ''),
                apiRequest.methodType,
                apiRequest.headers()
            );
            dispatch(storeSelectedIssuer(response?.response));
            setSelectedIssuer(response?.response);

            apiRequest = api.fetchIssuersConfiguration;
            response = await fetchRequest(
                apiRequest.url(params.issuerId ?? ''),
                apiRequest.methodType,
                apiRequest.headers()
            );
            dispatch(storeFilteredCredentials(response?.response));
            dispatch(storeCredentials(response?.response));
        };
        fetchCall();
    }, []);

    const previousPagePath = location.state?.from;

    const handleBackClick = () => {
        if (backUrl) {
            navigate(backUrl); // Navigate to the URL sent by the parent
        } else if (previousPagePath) {
            navigate(previousPagePath); // Navigate to the previous link in history
        } else {
            navigateToUserHome(navigate); // Navigate to homepage if opened directly
        }
    };

    return (
        <div
            data-testid={'Credential-Types-Page-Container'}
            className={CredentialTypesPageStyles.container}
        >
            <Header onBackClick={handleBackClick} displayObject={displayObject}
                    onClick={() => navigateToUserHome(navigate)}/>
            <CredentialTypesPageContent downloadStatus={downloadStatus} state={state}/>
        </div>
    );
};