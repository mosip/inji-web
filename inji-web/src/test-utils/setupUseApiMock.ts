import {RequestStatus} from "../utils/constants";

export const mockUseApi = {
    fetchData: jest.fn(),
    state: RequestStatus.LOADING,
    data: null as any,
    error: null as any,
    status: null,
    ok: () => false
};

type MockApiResponseOptions = {
    data?: object;
    status?: number | null;
    error?: any;
    headers?: object;
    state?: RequestStatus;
};

export function mockApiResponse({
                                    data = {},
                                    status = 200,
                                    error = null,
                                    headers = {},
                                    state = RequestStatus.DONE,
                                }: MockApiResponseOptions = {}) {
    mockUseApi.fetchData.mockResolvedValueOnce({
        ok: () => status === 200,
        data,
        status,
        error,
        headers,
        state
    });
}

export function mockApiResponseSequence(
    responses: MockApiResponseOptions[]
) {
    responses.forEach(
        (mockData) => {
            mockApiResponse(mockData)
        }
    );
}