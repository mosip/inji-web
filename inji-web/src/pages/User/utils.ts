import { User } from "../../components/Dashboard/types";
import { KEYS } from "../../utils/constants";
import {ROUTES} from "../../constants/Routes";

export const convertStringIntoPascalCase = (text: string | undefined) => {
    return (
        text?.toLocaleLowerCase()
            .split(' ')
            .map((word) => word.charAt(0).toUpperCase() + word.slice(1))
            .join(' ')
    );
};

export const navigateToUserHome = (navigate: any) =>
    navigate(ROUTES.USER_HOME);

export const validateWalletUnlockStatus = (
    cachedWalletId: string | null,
    storageWalletId: string | null,
    navigate: (path: string) => void,
    user: User
) => {
    if (cachedWalletId && (cachedWalletId === storageWalletId)) {
        console.info('Wallet is unlocked!');
    } else {
        console.warn(
            'Wallet exists but is locked, redirecting to `/pin` to unlock the wallet.'
        );
        if(user) {
            navigate('/pin');
            localStorage.removeItem(KEYS.WALLET_ID);
        }
    }
};