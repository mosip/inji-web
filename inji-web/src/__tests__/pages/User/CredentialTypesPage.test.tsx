import {fireEvent, screen, waitFor} from '@testing-library/react';
import {setMockUseDispatchReturnValue, setMockUseSelectorState} from '../../../test-utils/mockReactRedux';
import {CredentialTypesPage} from '../../../pages/User/CredentialTypes/CredentialTypesPage';
import {mockApiResponse, mockApiResponseSequence, mockUseApi} from "../../../test-utils/setupUseApiMock";
import {renderWithRouter} from "../../../test-utils/mockUtils";
import {credentialWellknown, mockCredentials, mockIssuerObjectList, userProfile} from "../../../test-utils/mockObjects";
import {api} from "../../../utils/api";
import {RequestStatus} from "../../../utils/constants";

const mockFetchUserProfile = jest.fn();
const mockIsUserLoggedIn = jest.fn();
let mockUseUserError: Error | null = null;

jest.mock('../../../hooks/useApi.ts', () => ({
    useApi: () => mockUseApi
}));

jest.mock('react-toastify', () => ({
    toast: {
        error: jest.fn(),
    },
}));

jest.mock('../../../hooks/User/useUser', () => ({
    useUser: () => ({
        fetchUserProfile: mockFetchUserProfile,
        isUserLoggedIn: mockIsUserLoggedIn,
        error: mockUseUserError
    })
}));

jest.mock('react-i18next', () => ({
    useTranslation: () => ({t: (key: string) => key}),
    initReactI18next: {
        type: '3rdParty',
        init: () => {
        },
    },
}));

const mockNavigate = jest.fn();
jest.mock('react-router-dom', () => ({
    ...jest.requireActual('react-router-dom'),
    useNavigate: () => mockNavigate,
}));

describe('CredentialTypesPage', () => {
    beforeEach(() => {
        jest.clearAllMocks();
        setMockUseDispatchReturnValue(jest.fn());
        mockApiResponseSequence([
            {response: {response: {display: [{language: 'en', name: 'Issuer1'}]}}},
            {response: credentialWellknown}
        ])

        mockFetchUserProfile.mockResolvedValue(userProfile)

        setMockUseSelectorState({
            credentials: {
                credentials: mockCredentials,
                filtered_credentials: mockCredentials,
            },
            issuers: {selected_issuer: mockIssuerObjectList[0] as any},
            common: {
                language: "en"
            }
        });
    });

    it('matches snapshot after data loads', () => {
        const {asFragment} = renderWithRouter(<CredentialTypesPage backUrl="/user/home"/>);

        expect(asFragment()).toMatchSnapshot();
    });

    it('renders CredentialTypesPage component', async () => {
        renderWithRouter(<CredentialTypesPage/>);

        await waitFor(() => expect(mockUseApi.fetchData).toHaveBeenCalledTimes(2));
        expect(mockUseApi.fetchData).toHaveBeenCalledWith({
            apiConfig: api.fetchSpecificIssuer,
            url: expect.stringContaining("/issuers")
        })
        expect(mockUseApi.fetchData).toHaveBeenCalledWith({
            apiConfig: api.fetchIssuersConfiguration,
            url: expect.stringContaining("/configuration")
        })
        expect(screen.getByTestId('credential-types-page-container')).toBeInTheDocument();
        await screen.findByText('Issuer1')
        await screen.findByText("Health Insurance")
    });

    it('should show error when fetchUserProfile fails', async () => {
        mockFetchUserProfile.mockRejectedValue(new Error('Error fetching user profile'));
        mockUseUserError = new Error("Error fetching user profile");


        renderWithRouter(<CredentialTypesPage/>);

        await waitFor(() => {
            expect(screen.getByTestId("EmptyList-Outer-Container")).toBeInTheDocument()
        });
    });

    it('should show error when fetching issuer fails', async () => {
        mockUseApi.fetchData.mockReset()
        // mockUseApi.fetchData.mockRejectedValueOnce(new Error('Error fetching issuer'));
        mockApiResponse({
            error: new Error("Error fetching issuer"),
            status: 400,
            state: RequestStatus.ERROR
        })
        // mockUseApi.state = RequestStatus.ERROR

        renderWithRouter(<CredentialTypesPage/>);

        await waitFor(() => expect(mockUseApi.fetchData).toHaveBeenCalledWith({
            apiConfig: api.fetchSpecificIssuer,
            url: expect.stringContaining("/issuers")
        }));
        expect(screen.getByText("emptyContainerContent")).toBeInTheDocument();
    });

    it.todo('should show error when fetching issuer configuration fails');

    it.todo('should navigate to credentials page when download is successful');

    it.todo('cleans up download session IDs on unmount');

    it('should navigate to home when Home button is clicked', () => {
        renderWithRouter(<CredentialTypesPage backUrl={"/previous-url"}/>);

        fireEvent.click(screen.getByRole("button", {name: "Common:home"}))

        expect(mockNavigate).toHaveBeenCalledWith("/user/home");
    });

    it('should navigate to previous path when available and back button is clicked', () => {
        const previousPath = '/previous-path';
        renderWithRouter(<CredentialTypesPage backUrl={previousPath}/>);

        fireEvent.click(screen.getByTestId("back-arrow-icon"));

        expect(mockNavigate).toHaveBeenCalledWith(previousPath);
    })

    it.todo('should show loader while downloading card')
    it('should navigate to home when back button is clicked & previous path not available', () => {
        renderWithRouter(<CredentialTypesPage/>);

        fireEvent.click(screen.getByTestId("back-arrow-icon"));

        expect(mockNavigate).toHaveBeenCalledWith("/user/home");
    })
});
