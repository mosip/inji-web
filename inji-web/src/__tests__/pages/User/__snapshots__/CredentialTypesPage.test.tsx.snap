// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CredentialTypesPage matches snapshot after data loads 1`] = `
<DocumentFragment>
  <div
    class="container mx-auto sm:px-2 md:px-4 lg:px-6 py-6 flex flex-col relative gap-4 sm:gap-6 md:gap-10 lg:gap-12 ml-4 sm:ml-0 h-screen"
    data-testid="credential-types-page-container"
  >
    <div
      class="flex flex-col lg:flex-row justify-between items-start gap-4 sm:gap-2"
    >
      <div
        class="flex items-start"
      >
        <div
          class="flex items-start"
        >
          <svg
            class="mr-2 cursor-pointer"
            data-testid="back-arrow-icon"
            fill="none"
            height="29"
            viewBox="0 0 24 18"
            width="29"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M20 11H7.83l5.59-5.59L12 4l-8 8 8 8 1.41-1.41L7.83 13H20v-2z"
              fill="#000000"
            />
          </svg>
        </div>
        <div
          class="flex flex-col items-start"
        >
          <span
            class="text-2xl font-medium"
            data-testid="Stored-Credentials"
          />
          <button
            class="text-xs sm:text-sm text-iw-secondary cursor-pointer hover:underline "
            data-testid="btn-home"
          >
            Common:home
          </button>
        </div>
      </div>
      <div>
        <div
          class="flex items-center w-full justify-start sm:justify-end my-5 sm:my-0"
          data-testid="NavBar-Search-Container"
        >
          <div
            class="w-full sm:w-96 flex justify-center items-center bg-iw-background shadow-iw rounded-md"
            data-testid="Search-Issuer-Container"
          >
            <svg
              class="m-5"
              color="var(--iw-color-searchIcon)"
              data-testid="NavBar-Search-Icon"
              fill="currentColor"
              height="22"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 512 512"
              width="22"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M505 442.7L405.3 343c-4.5-4.5-10.6-7-17-7H372c27.6-35.3 44-79.7 44-128C416 93.1 322.9 0 208 0S0 93.1 0 208s93.1 208 208 208c48.3 0 92.7-16.4 128-44v16.3c0 6.4 2.5 12.5 7 17l99.7 99.7c9.4 9.4 24.6 9.4 33.9 0l28.3-28.3c9.4-9.4 9.4-24.6.1-34zM208 336c-70.7 0-128-57.2-128-128 0-70.7 57.2-128 128-128 70.7 0 128 57.2 128 128 0 70.7-57.2 128-128 128z"
              />
            </svg>
            <input
              class="py-6 w-11/12 flex text-iw-searchTitle focus:outline-none overflow-ellipsis mr-10"
              data-testid="NavBar-Search-Input"
              placeholder="searchText"
              type="text"
              value=""
            />
          </div>
        </div>
      </div>
    </div>
    <div>
      <div
        data-testid="Credential-List-Container"
      >
        <div
          class="flex justify-center items-center"
          data-testid="SpinningLoader-Container"
        >
          <div
            aria-busy="true"
            aria-label="oval-loading"
            class="sc-beySPh gZmPOF"
            data-testid="oval-loading"
            role="progressbar"
          >
            <svg
              data-testid="oval-svg"
              height="80"
              stroke="var(--iw-color-spinningLoaderPrimary)"
              viewBox="-20.5 -20.5 43 43"
              width="80"
              xmlns="http://www.w3.org/2000/svg"
            >
              <g
                fill="none"
                fill-rule="evenodd"
              >
                <g
                  data-testid="oval-secondary-group"
                  stroke-width="3"
                  transform="translate(1 1)"
                >
                  <circle
                    cx="0"
                    cy="0"
                    r="20"
                    stroke="var(--iw-color-spinningLoaderSecondary)"
                    stroke-opacity=".5"
                    stroke-width="3"
                  />
                  <path
                    d="M20 0c0-9.94-8.06-20-20-20"
                  >
                    <animatetransform
                      attributeName="transform"
                      dur="1s"
                      from="0 0 0"
                      repeatCount="indefinite"
                      to="360 0 0"
                      type="rotate"
                    />
                  </path>
                </g>
              </g>
            </svg>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
